atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManager.java

```diff
@@ -224,35 +236,63 @@ public List<OpenTransaction> startTransactions(List<? extends PreCommitCondition
             recordImmutableTimestamp(immutableTs);
             cleaner.punch(responses.get(0).startTimestampAndPartition().timestamp());
 
-            List<OpenTransaction> transactions = Streams.zip(
+            List<OpenTransactionImpl> transactions = Streams.zip(
                             responses.stream(), conditions.stream(), (response, condition) -> {
                                 LockToken immutableTsLock =
                                         response.immutableTimestamp().getLock();
                                 LongSupplier startTimestampSupplier = response.startTimestampAndPartition()::timestamp;
 
                                 ExpectationsAwareTransaction transaction = createTransaction(
                                         immutableTs, startTimestampSupplier, immutableTsLock, condition);
-                                transaction.onSuccess(
-                                        () -> lockWatchManager.onTransactionCommit(transaction.getTimestamp()));
+
+                                transaction.onCommitOrAbort(transaction::reportExpectationsCollectedData);
+                                transaction.onCommitOrAbort(condition::cleanup);
+                                transaction.onCommitOrAbort(
+                                        () -> lockWatchManager.requestTransactionStateRemovalFromCache(
+                                                response.startTimestampAndPartition()
+                                                        .timestamp()));
+                                // N.B. run this before clearing the state from the cache with
+                                // requestTransactionStateRemovalFromCache
```

Comment:
This was quite nuanced:
- `lockWatchManager.onTransactionCommit` needs to run before `lockWatchManager.requestTransactionStateRemovalFromCache`
- onSuccess ran before the finally block before, and therefore it worked
- now we need to add to `onCommitOrAbort` in the right order, otherwise tests failed.

Reply:
Also thought of writing
```
                                transaction.onCommitOrAbort(() -> {
                                    if (!transaction.isAborted()) {
                                        // in onCommitOrAbort + not aborted == committed
                                        lockWatchManager.onTransactionCommit(transaction.getTimestamp());
                                    }
                                    lockWatchManager.requestTransactionStateRemovalFromCache(....)
                                });
```
but decided not do this because I like guaranteeing both are run in case one fails.

Reply:
Also thought of adding `txn.isCommitted()` as a method, but not doing that for now - could pick do in a flup pr if we decide to go down that route.